### PR TITLE
QuerySpecs creation with targetResourceType or targetResource Class

### PR DESCRIPTION
**Pull request by jbmolle**
Hi @remmeier 
This PR should solve #636 
I've split relatedSpecs between resourceType and resourceClass as you proposed.
I've used a Set to return the relatedSpecs in order to remove duplicates Specs (those who are both related to a resourceType and a resourceClass).
Let me know if it is what you were looking for.
Thx
JB

### CODE REVIEW

**Comment by jbmolle** [comment]
Hi,
I think that we shouldn't allow QuerySpec to have null resourceType.
On the second commit, in QuerySpec if the resourceType is null then I get it from the resourceClass through the JsonApiResource annotation.

We have actually 2 cases:
- Either there is one resource class for each resource type. So we can map the resource type to the resource class (through JsonApiResource annotation)
- Or there is one resource class for multiple resource types. For example if we are creating resources dynamically at runtime. Then when we create a QuerySpec, we must specify the resource type. Just giving the resource class won't be enough to know which resources it corresponds to.

Currently if we have multiple resource types for one resource class, the mappedBy relations won't work. The ForwardOppositeStrategy repository won't build the correct queryspecs and the application return an error 500.
This PR also solves this problem.

**Comment by jbmolle** [comment]
Hi @remmeier 
Did you have time to review this PR?
Let me know if I need to change something.
Thanks!

File: crnk-core/src/main/java/io/crnk/core/queryspec/QuerySpec.java [hunk 1]
```diff
@@ -261,16 +266,20 @@ public void setIncludedRelations(List<IncludeRelationSpec> includedRelations) {
...
+    	allRelatedSpecs.addAll(classRelatedSpecs.values());
+        return Collections.unmodifiableCollection(allRelatedSpecs);
     }
 
     public void setNestedSpecs(Collection<QuerySpec> specs) {
-        this.relatedSpecs.clear();
+    	this.typeRelatedSpecs.clear();;
+        this.classRelatedSpecs.clear();
         for (QuerySpec spec : specs) {
             if (spec.getResourceClass() != null) {
```
Comment (remmeier) [inline crnk-core/src/main/java/io/crnk/core/queryspec/QuerySpec.java]: I suspect here the getResourceType() should have priority?

Comment (jbmolle) [reply crnk-core/src/main/java/io/crnk/core/queryspec/QuerySpec.java]: Yes you're right. I've just updated the pull request to reflect the changes

**Comment by remmeier** [comment]
does .`/gradlew build` work for you? it seems that the travis integration is not working just now. Have to take a look at that. If so we could merge it.

**Comment by jbmolle** [comment]
Now it should work.
Actually in setNestedSpecs we shouldn't do first resourceType and then resourceClass. We add it to both typeRelatedSpecs and classRelatedSpecs.
When doing getNestedSpecs, we send back a Set so QuerySpecs both in typeRelatedSpecs and classRelatedSpecs won't be sent twice.

The problem came from setNestedSpecWithClass() test:
we do Assert.assertSame(relatedSpec, spec.getQuerySpec(Task.class)); but in setNestedSpecs we were adding it only to typeRelatedSpecs so it wasn't in classRelatedSpecs.
The test Assert.assertSame(relatedSpec, spec.getQuerySpec("tasks")); was working.
Sorry about that.

**Comment by remmeier** [comment]
perfect thx! a crnk 3.2 should be available within the coming days in case you are on latest.

**Pull request merged**